File: javafx/src/com/mku/salmon/vault/controller/PDFViewerController.java
package com.mku.salmon.vault.controller;
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import com.mku.salmon.vault.dialog.SalmonDialog;
import com.mku.salmon.vault.viewmodel.SalmonFileViewModel;
import javafx.scene.control.Alert;
import javafx.stage.Stage;
import org.icepdf.ri.common.ComponentKeyBinding;
import org.icepdf.ri.common.SwingController;
import org.icepdf.ri.common.SwingViewBuilder;

import javax.swing.*;
import java.io.IOException;
import java.io.InputStream;

public class PDFViewerController {

    public static void openPDFViewer(SalmonFileViewModel file, Stage owner) throws IOException {
        SwingController swingController = new SwingController();
        SwingViewBuilder factory = new SwingViewBuilder(swingController);
        JPanel viewerComponentPanel = factory.buildViewerPanel();
        ComponentKeyBinding.install(swingController, viewerComponentPanel);
        swingController.getDocumentViewController().setAnnotationCallback(
                new org.icepdf.ri.common.MyAnnotationCallback(
                        swingController.getDocumentViewController()));
        JFrame window = new JFrame("PDF Viewer");
        window.getContentPane().add(viewerComponentPanel);
        window.pack();
        window.setVisible(true);
        try {
            InputStream stream = file.getAesFile().getInputStream().asReadStream();
            swingController.openDocument(stream, "Encrypted PDF", file.getAesFile().getName());
        } catch (Exception e) {
            e.printStackTrace();
            new SalmonDialog(Alert.AlertType.ERROR, "Could not load PDF: " + e.getMessage()).show();
        }
    }
}

File: js/src/assets/js/vault/controller/pdf_viewer_controller.js
/*
MIT License

Copyright (c) 2025 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/


import { JWindow } from "../../lib/jwin/assets/js/jwindow.js";
import { JBind } from "../../lib/jbind/jbind.js";
import { ObjectProperty } from "../../lib/jbind/object_property.js";
import { BooleanProperty } from "../../lib/jbind/boolean_property.js";
import { MemoryStream } from "../../lib/simple-io/streams/memory_stream.js";
import { Handler } from "../../lib/salmon-fs/service/handler.js";

export class PdfViewerController {
    static contentURL = "pdf-viewer.html";
    iframe;
    contentWindow;
    viewer;
    progressVisibility;
    url;

    setStage(contentWindow) {
        this.contentWindow = contentWindow;
        this.iframe = JBind.bind(this.contentWindow.getRoot(), 'pdf-viewer-iframe', 'src', new ObjectProperty());
        this.progressVisibility = JBind.bind(this.contentWindow.getRoot(), 'pdf-progress', 'display', new BooleanProperty());
    }

    static async openPdfViewer(fileViewModel, owner) {
        let controller = new PdfViewerController();
        let contentWindow = await JWindow.createWindowWithURL("PDF Viewer", this.contentURL);
        contentWindow.modal.style.resize = "both";
        contentWindow.modal.style.width = "800px";
        contentWindow.modal.style.height = "600px";
        controller.setStage(contentWindow);
        setTimeout(() => {
            controller.load(fileViewModel);
        });
        contentWindow.show();
        contentWindow.onClose = () => controller.onClose(this);
    }

    async load(fileViewModel) {
        try {
            let stream = await fileViewModel.getAesFile().getInputStream();
            let ms = new MemoryStream();
            await stream.copyTo(ms);
            await stream.close();
            let blob = new Blob([ms.toArray().buffer], { type: "application/pdf" });
            await ms.close();
            this.url = URL.createObjectURL(blob);
            this.iframe.set(this.url);
        } catch (e) {
            console.error(e);
        }
        this.progressVisibility.set(false);
    }

    onClose() {
        URL.revokeObjectURL(this.url);
        Handler.getInstance().unregister();
    }
}
